In [1]:
!pip install bertopic sentence-transformers umap-learn hdbscan spacy nltk
!python -m spacy download ru_core_news_sm

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 150.6/150.6 kB 1.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 26.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 26.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 16.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 10.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 32.5 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstalli

In [11]:
import pandas as pd
import re
import spacy
import nltk
from nltk.corpus import stopwords
from tqdm.notebook import tqdm
from sklearn.feature_extraction.text import CountVectorizer
from bertopic import BERTopic
from umap import UMAP
import hdbscan
from sentence_transformers import SentenceTransformer
from collections import Counter
import numpy as np

In [3]:
nltk.download('stopwords')
russian_stopwords = stopwords.words("russian")


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [4]:
nlp_spacy = spacy.load("ru_core_news_sm")

def clean_text(text):
    text = re.sub(r'\s+', ' ', text)
    text = re.sub(r'[^а-яА-ЯёЁ ]', '', text)
    return text.strip().lower()

def preprocess_spacy(text):
    text = clean_text(text)
    doc = nlp_spacy(text)
    lemmas = [token.lemma_ for token in doc if token.lemma_ not in russian_stopwords and token.is_alpha]
    return ' '.join(lemmas)


In [5]:
!wget https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.1/lenta-ru-news.csv.bz2

--2025-05-01 12:08:11--  https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.1/lenta-ru-news.csv.bz2
Resolving github.com (github.com)... 140.82.112.4
Connecting to github.com (github.com)|140.82.112.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/619f9f00-1e96-11ea-946e-dac89df8aced?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250501%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250501T120811Z&X-Amz-Expires=300&X-Amz-Signature=d85db89d1e6b145bf2073913af6cc373ab7536682a31c6ca5c1f04f91883d62d&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dlenta-ru-news.csv.bz2&response-content-type=application%2Foctet-stream [following]
--2025-05-01 12:08:11--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/619f9f00-1e96-11ea-946e-dac89df8aced?X-Amz-Algorithm=AWS4-HMA

In [6]:

raw_df = pd.read_csv("lenta-ru-news.csv.bz2")
df = raw_df[raw_df['text'].notnull()].sample(n=5000, random_state=42)

preprocessed_texts = []
for text in tqdm(df['text']):
    preprocessed_texts.append(preprocess_spacy(text))

<ipython-input-6-95ac6b885c4d>:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  raw_df = pd.read_csv("lenta-ru-news.csv.bz2")


  0%|          | 0/5000 [00:00<?, ?it/s]

In [7]:
embedding_model = SentenceTransformer('paraphrase-multilingual-MiniLM-L12-v2')
umap_model = UMAP(n_neighbors=15, n_components=5, min_dist=0.0, metric='cosine', random_state=42)
hdbscan_model = hdbscan.HDBSCAN(min_cluster_size=15, metric='euclidean', cluster_selection_method='eom', prediction_data=True)
vectorizer_model = CountVectorizer(ngram_range=(1, 2), stop_words=russian_stopwords)

topic_model = BERTopic(
    embedding_model=embedding_model,
    umap_model=umap_model,
    hdbscan_model=hdbscan_model,
    vectorizer_model=vectorizer_model,
    language='russian',
    calculate_probabilities=True,
    verbose=True
)

modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/3.89k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/645 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/471M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/480 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


tokenizer.json:   0%|          | 0.00/9.08M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [8]:
topics, probs = topic_model.fit_transform(preprocessed_texts)

2025-05-01 12:21:16,034 - BERTopic - Embedding - Transforming documents to embeddings.


Batches:   0%|          | 0/157 [00:00<?, ?it/s]

2025-05-01 12:31:28,121 - BERTopic - Embedding - Completed ✓
2025-05-01 12:31:28,126 - BERTopic - Dimensionality - Fitting the dimensionality reduction algorithm
2025-05-01 12:32:12,246 - BERTopic - Dimensionality - Completed ✓
2025-05-01 12:32:12,250 - BERTopic - Cluster - Start clustering the reduced embeddings
2025-05-01 12:32:13,875 - BERTopic - Cluster - Completed ✓
2025-05-01 12:32:13,900 - BERTopic - Representation - Fine-tuning topics using representation models.
2025-05-01 12:32:19,136 - BERTopic - Representation - Completed ✓


In [9]:
topic_model.visualize_topics()
topic_model.visualize_documents(preprocessed_texts, topics=topics)
topic_model.visualize_distribution(probs[0], min_probability=0.015)

In [10]:
def topic_diversity(topics, top_n=10):
    unique_words = set()
    total_words = 0
    for topic in topics:
        words = topic_model.get_topic(topic)
        if words is None:
            continue
        for word, _ in words[:top_n]:
            unique_words.add(word)
            total_words += 1
    return len(unique_words) / total_words

all_topic_ids = topic_model.get_topic_info()['Topic'].tolist()
all_topic_ids = [tid for tid in all_topic_ids if tid != -1]
diversity_score = topic_diversity(all_topic_ids, top_n=10)
print(f"Topic Diversity: {diversity_score:.4f}")

Topic Diversity: 0.7783


In [13]:
def build_df_stats(tokenized_docs):
    doc_counts = Counter()
    pair_counts = Counter()
    for doc in tokenized_docs:
        unique_words = set(doc)
        for w in unique_words:
            doc_counts[w] += 1
        for i in range(len(doc)):
            for j in range(i + 1, len(doc)):
                w1, w2 = sorted((doc[i], doc[j]))
                pair_counts[(w1, w2)] += 1
    return doc_counts, pair_counts

tokenized = [doc.split() for doc in preprocessed_texts]
doc_freq, pair_freq = build_df_stats(tokenized)

def umass_score_for_topic(topic_words, doc_freq, pair_freq, total_docs):
    score = 0
    pairs = 0
    for i in range(1, len(topic_words)):
        for j in range(i):
            wi, wj = topic_words[i], topic_words[j]
            pair = tuple(sorted((wi, wj)))
            D_wj = doc_freq.get(wj, 1)
            D_wi_wj = pair_freq.get(pair, 0)
            score += np.log((D_wi_wj + 1) / D_wj)
            pairs += 1
    return score / pairs if pairs > 0 else 0

topics_info = topic_model.get_topic_info()
valid_topics = topics_info[topics_info.Topic != -1]['Topic'].tolist()
topic_umasses = []
for topic_id in valid_topics:
    words = [word for word, _ in topic_model.get_topic(topic_id)[:10]]
    score = umass_score_for_topic(words, doc_freq, pair_freq, len(tokenized))
    topic_umasses.append(score)

avg_umass = np.mean(topic_umasses)
print(f"Средний UMass: {avg_umass:.4f}")

Средний UMass: -0.7686


Проведена предобработка с использованием spaCy:

Построена BERTopic-модель с кастомным пайплайном (SentenceTransformer + UMAP + HDBSCAN + CountVectorizer), адаптированным под русский язык.

Проведены визуализации:

ключевые слова в темах,

проекция документов в 2D,

распределение вероятностей тем.

Вычислены метрики качества:

Topic Diversity = 0.7783 — высокий показатель разнообразия, темы содержат уникальные токены;

UMass Coherence = –0.7686 — приемлемый уровень когерентности, типичный для несемейных BERTopic моделей без дообучения.

Проблемы:
Некоторые темы оказались «размытыми» или дублирующимися — модель HDBSCAN склонна к созданию мелких, слабосвязанных кластеров.

Отсутствие ярко выраженных заголовков может снижать когерентность тем.

Ограничение на размер обучающей выборки (5000 документов) влияет на полноту тем.

Улучшение:
Попробовать другие sentence embedding модели

Использовать альтернативную кластеризацию

Протестировать модель на большем объёме данных